library/README.md

```diff
@@ -119,35 +125,90 @@ NDK_HOME=$HOME/Documents/GitHub/engine/src/third_party/android_tools/ndk cargo n
 - Updates must be applied in order.
 - Updates are applied in a single transaction.
 
-### Update State Machine
+### State Machines
+
+#### Boot State Machine
+
+This state machine tracks the process of the engine starting up, with or without
+a patch. These state transitions happen whether or not a new patch is available,
+but in the context of the updater, we only care about the case where we are
+booting from a patch.
+
+The patch boot state is internal to the PatchManager and stored on disk in
+`patches_state.json`. It contains three fields, all of which are Optional
+PatchMetadata structs:
+
+- last_boot_patch: The last patch that was successfully booted.
+- last_attempted_patch: The last patch that we attempted to boot.
+- next_boot_patch: The next patch that we will attempt to boot.
+
+This state machine has the following states. It can only move forward through
+them.
+
+1. Ready - The engine is initialized but has not started booting yet.
+2. Booting - The engine has started booting.
+3. Success - The engine successfully booted.
+4. Failure - The engine failed to boot.
+
+State is advanced through calls to the following methods of the `ManagePatches`
+trait, which PatchManager implements:
+
+- `record_launch_start`: Moves from Ready to Booting
+  - next_boot_patch is the patch that we will attempt to boot, i.e., the "current" patch.
+  - last_attempted_patch is set to next_boot_patch.
+- `record_launch_success`: Moves from Booting to Success
+  - last_boot_patch is set to next_boot_patch.
+  - Artifacts for patches older than last_boot_patch are deleted.
+- `record_launch_failure`: Moves from Booting to Failure
+  - next_boot_patch artifacts are deleted.
+  - next_boot_patch is set to either:
+    - last_boot_patch if it is still valid, or
+    - None, if last_boot_patch is None or invalid.
+
+These are effectively no-ops if we are not booting from a patch.
+
+We assume that this state machine will have advanced at least as far as the
+Booting state before the Patch Check State Machine (below) is started.
```

Comment:
Is any call to an earlier method an error?  Should it be?

Reply:
This also became part of the "not enforced but should be" assumptions list above.